network/discovery.go

```diff
@@ -259,7 +259,7 @@ func (d *discovery) run() {
 
 func (d *discovery) handleDiscovery() {
 	// take a random peer and find peers
-	if target := d.peers.getRandomPeer(); target != nil {
+	if target := d.peers.getRandomPeer(); d.srv.numOpenSlots() > 0 && target != nil {
```

Comment:
Why would you get a random peer if you don't even check that you can dial them beforehand, but you check after you grabbed a random peer?

getRandomPeer is not a cheap operation as it involves locking and spinning up the random number generator

Reply:
If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/7fa13584e2cdc484c1189865c7b2a51bc7ecec5c